Hotrod the Beetle - main parts complete

NecroReindeer opened the pull request:
I have finished the biggest jobs  - level generation, character movement and enemy decision-making.
Everything look ok so far? Are my sprint plans in the README.md of the main folder ok?

(Also, has a peer review of my work from the session before Christmas been done?)

NecroReindeer committed:
Uploaded screenshot with user stories

NecroReindeer committed:
Made a circle that moves in the direction of swipes

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Generate grid of cells and walls

The program now generates a grid of cells which each have 4 walls

NecroReindeer committed:
Grid is now centred

NecroReindeer committed:
PlayArea and Wall now defined in .kv file

NecroReindeer committed:
Tidied up code. Added enum for direction

NecroReindeer committed:
Added CellEdge class to store the type of edge

Added CellEdge that can be a wall or passage.
Added enum for wall and passage.
Added method in direction enum for converting directions to angles.

NecroReindeer committed:
Changed cells to be stored in 2-dimensional list

NecroReindeer committed:
Added function to track player position on grid

NecroReindeer committed:
Implemented maze generation algorithm

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Tidied up code

Put cell classes into a separate file. Restructured code. Put level
generation in a separate level class. Moved movement detection to root
HotrodGame.

NecroReindeer committed:
Can now queue directions and not go through walls

NecroReindeer committed:
Can now queue moves and not go through walls

NecroReindeer committed:
Implemented algorithm to remove dead ends

Implemented algorithm to remove dead ends.
Also fixed the cell list - rows and columns were the wrong way round

NecroReindeer committed:
Added pellets that can be collected to increase score

Added pellets that can be collected to increase score.
Moved PlayerBeetle widget to a character.py.
Moved Level widget to level.py.

NecroReindeer committed:
Added screenshots of Trello Board sprint plans

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Speed now calculated relative to cell_size

NecroReindeer committed:
Restructured code

Main game elements (i.e. level, player, enemy) can now access each other
through the HotrodGame. HotrodGame now has access to all main game
elements.
Made PlayerBeetle movement methods part of a parent Character class, and
made PlayerBeetle a subclass.
Also subclassed EnemyBeetle.

NecroReindeer committed:
Create README.md

NecroReindeer committed:
Added one enemy

NecroReindeer committed:
Fixed bug that caused some dead ends to remain

NecroReindeer committed:
Added pink enemy and its movement

NecroReindeer committed:
Tidied up code. Restructured and added comments.

Added methods in level for converting between grid coordinates and
window coordinates.
Added comments and docstrings.
Removed some instances of repeating code using loops and variables.
Changed
Also added visualisation of enemy targets (for testing purposes).

NecroReindeer committed:
Added blue enemy and implemented movement

NecroReindeer committed:
Added orange enemy and changed how speed works

The default speed is now moving 1 cell per second, calculated using the
frame rate and cell size. This value is then multiplied by a speed
multiplier, which can be used to adjust the speed.

NecroReindeer committed:
Added screenshot of trello board for third sprint

NecroReindeer committed:
Added third sprint screenshot

NecroReindeer committed:
Added lives and checks for colliding with enemies

Lives are now tracked. If character is on same grid space as an enemy, a
life is lost and character positions are reset.

NecroReindeer committed:
A new level is now generated upon losing all lives

NecroReindeer committed:
Fixed crash when window size changes

Size and position of all objects in the play area are now readjusted
when the level widget's size changes (via a kivy event)

NecroReindeer committed:
Added game over screen and fixed bugs

A game over screen is now displayed upon losing all lives
Also, fixed crash when window size changes (previous commit that said
that was the wrong files...)

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Added scatter mode and changed enemy movement

Enemies now switch between chasing and scattering.
When deciding their next direction, the selection is now prioritised
up-left-down-right if more than one are equal distance

NecroReindeer committed:
Added fourth sprint screenshot

NecroReindeer committed:
Update README.md

NecroReindeer committed:
Update README.md

Adrir commented:
Sprint plans look great. If you do not receive a review from your peers by the end of the week, then send me an email.

NecroReindeer committed:
Added method create enemy den. Fixed bug.

Added method to create the walls of the enemy den and set passages
between the surrounding cells.
Fixed bug where enemy mode change would be scheduled multiple times upon
death.

Adrir commented:
Code looks great. 

Only a few nit picks that I have noticed skimming through and things you could do to go the extra mile. There seem to be a few arbitrary literals defined throughout the code. These could be declared as constants (note: a constant in Python is just a variable that does not change) and/or moved into a different file (e.g. the .kv file, .ini file, etc.) to improve maintainability. I would also take the opportunity to make the doc-strings more verbose, where appropriate, and check them against a [validation tool](http://dolphm.com/pep257-good-python-docstrings-by-example/). If you want to go even further, then consider adding a [CHANGELOG.md](https://github.com/skywinder/Github-Changelog-Generator/blob/master/CHANGELOG.md) to your project.
